google/cloud/bigtable/internal/rate_limiter.h

```diff
@@ -0,0 +1,118 @@
+// Copyright 2023 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_BIGTABLE_INTERNAL_RATE_LIMITER_H
+
+#include "google/cloud/internal/clock.h"
+#include "google/cloud/log.h"
+#include "google/cloud/version.h"
+#include <chrono>
+#include <cstdint>
+#include <mutex>
+
+namespace google {
+namespace cloud {
+namespace bigtable_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+/**
+ * A threadsafe interface for rate limiting.
+ *
+ * The caller tells the `RateLimiter` how many permits it wants to acquire. The
```

Comment:
I would start with something like:

```
The caller needs to acquire a "permit" to perform the operation under rate limits. This class limits the number of
permits issued per period of time, effectively limiting the operation rate.
The caller may acquire more than one permit at a time, if it needs to perform a burst of the operation under rate limit.
More permits become available as time passes, with some maximum to limit the size of bursts.
```

I think "token" or "credit" is a more common name than "permit", but is not like any of the three is universally accepted.  

You may want to quote: https://en.wikipedia.org/wiki/Flow_control_(data)#Open-loop_flow_control or whatever ATM networks called their flow control mechanism.

Reply:
> I would start with something like:

Thanks, it is what I was trying to say, but clearer.

> or whatever ATM networks called their flow control mechanism.

Why do you know these things?